### PR TITLE
Some good changes in Job module

### PR DESCRIPTION
**Pull request by tallysmartins**
- Support to wait-for, added to `command` of the docker-compose.yml
- Remove `run_times` from measurements results
- Refactor Job module and add some tests

### CODE REVIEW

**Comment by tallysmartins** [comment]
@PragTob work is getting big here, can you take a look and give me some feedback? We dindn't talk to much about what I'm doing here, so just to clarify things for you:

 1. My initial intention was just change the `process_job` function in the Runner module to send the results of the benchmarks each one per request and write the proper tests for it
 2. I've found that testing this would need changes in the way we connect to the API, something similar to what I did in the server-api, using different Clients, so I left this changes for later
 3. I've found that there were no tests setup yet
 4. I've added tests setup and started testing the Job module 
 5. I've found that it need refactoring because was very hard to test the complex processing in the uniq public method `start_job`, so I changed it to receive  a "processing function"  to become easier to test
 6. I also started to make functions public and add `@doc false` to test them. I've read  [this post from PragDave](https://pragdave.me/blog/2017/03/31/tesing-private-functions.html) about testing private functions.

My plan is to finish testing the functions of the Job module and then work on testing the Runner module, which is a GenServer and contain the logic to send the results to the api. What do you think?

**Comment by PragTob** [comment]
Sorry I seem to have missed this one not getting a notification for it :'(

File: lib/runner.ex [hunk 1]
```diff
@@ -36,17 +37,11 @@ defmodule ElixirBench.Runner do
...
-      "commit_sha" => commit,
-      "config" => config
-    } = job
+  defp process_job(job_data, client) do
+    opts = [timeout: @job_timeout, task_fun: &Job.run_job/1]
+    job = build_job_struct(job_data)
 
-    config = Config.from_string_map(config)
-    job = Job.start_job(id, repo_slug, branch, commit, config)
+    job = Job.start_job(job, opts)
```
Comment (PragTob) [inline lib/runner.ex]: seems like we're missing opportunity to use the great pipe here :) `|>`

File: lib/runner.ex [hunk 2]
```diff
@@ -58,4 +53,24 @@ defmodule ElixirBench.Runner do
...
+
+    config = Config.from_string_map(config)
+
+    %Job{
+      id: to_string(id),
+      repo_slug: repo_slug,
+      branch: branch,
+      commit: commit,
+      config: config
+    }
```
Comment (PragTob) [inline lib/runner.ex:73]: it's not immedeatly obvious what the benefit of building the job struct here is. Not sure it should be or how to make it better. Just an observation. I think we do it so that we have a job struct to work with rather than the data map. Then it's also weird though why some of the fields are called slightly different - we're in control of all those field names right? Or am I missing something? Then we should probably call them all the same, then creating a job struct from it is also much easier.

File: lib/runner/job.ex [hunk 3]
```diff
@@ -36,24 +36,17 @@ defmodule ElixirBench.Runner.Job do
...
+  def start_job(job, opts) do
     ensure_no_other_jobs!()
 
-    job = %Job{
-      id: to_string(id),
-      repo_slug: repo_slug,
-      branch: branch,
-      commit: commit,
-      config: config
-    }
```
Comment (PragTob) [inline lib/runner/job.ex:48]: I see you moved the responsibility. I think that's a good thing, that the job module just deals with job structs here so I like it :+1:

File: lib/runner/job.ex [hunk 4]
```diff
@@ -125,11 +120,26 @@ defmodule ElixirBench.Runner.Job do
...
-      environment: build_runner_environment(job)
+      depends_on: Map.keys(deps),
+      environment: build_runner_environment(job),
+      command: build_runner_command_from_deps(deps)
     }
   end
 
+  defp build_runner_command_from_deps(deps) do
+    command =
+      Enum.reduce(deps, "", fn {_dep_name, dep_params}, acc ->
```
Comment (PragTob) [inline lib/runner/job.ex]: I think renaming `acc` to something like `command` here helps the readability as it makes it more obvious the accumulator is hopening the command or maybe the `wait_command` something like that

File: lib/runner/job.ex [hunk 5]
```diff
@@ -36,24 +36,17 @@ defmodule ElixirBench.Runner.Job do
...
 
-    job = %Job{
-      id: to_string(id),
-      repo_slug: repo_slug,
-      branch: branch,
-      commit: commit,
-      config: config
-    }
+    timeout = Keyword.get(opts, :timeout, Confex.fetch_env!(:runner, :job_timeout))
+    task_fun = Keyword.get(opts, :task_fun, &run_job/1)
```
Comment (PragTob) [inline lib/runner/job.ex]: not sure if having the task_function be part of the options is the best solution here. Having it be part of the options sort of suggests to me that it's viable to pass something different in there. In reality we just want to have it for testing purposes.

I think using a default argument like `runner_function \\ &fun` works in our favor here

File: test/runner/job_test.exs [hunk 6]
```diff
@@ -45,4 +19,253 @@ defmodule ElixirBench.Runner.JobTest do
...
+      end
+
+      %{pid: pid} =
+        Task.Supervisor.async_nolink(ElixirBench.Runner.JobsSupervisor, fn ->
+          :timer.sleep(2_000)
+        end)
+
+      assert_raise MatchError, fn ->
+        Job.start_job(job_fixture(), task_fun: task_fun)
+      end
```
Comment (PragTob) [inline test/runner/job_test.exs]: I'm not sure how this works? :thinking: We only launch one job here, why should it fail? I guess it's in the supervisor above but at least to me that connection isn't clear

File: test/runner/job_test.exs [hunk 7]
```diff
@@ -45,4 +19,253 @@ defmodule ElixirBench.Runner.JobTest do
...
+
+      job = Job.start_job(job_fixture(), task_fun: task_fun, timeout: 100)
+
+      assert %{log: "Job execution timed out", status: 127} = job
+    end
+  end
+
+  describe "ensure_no_other_jobs!/0" do
+    test "raise error if active child process already exist under JobsSupervisor" do
```
Comment (PragTob) [inline test/runner/job_test.exs]: this might be controversial, as I don't think anyone else calls it I think it should be tested as part of `start_job` and this is just an implementation detail and keep the method private.

That gives us a smaller API surface and less possible rework due to API changes as we could completely remove the method as long as we keep the functionality

File: lib/runner/job.ex [hunk 8]
```diff
@@ -93,11 +86,13 @@ defmodule ElixirBench.Runner.Job do
...
-  defp get_benchmars_output_path(%Job{id: id}) do
+  @doc false
+  def get_benchmars_output_path(%Job{id: id}) do
     Confex.fetch_env!(:runner, :benchmars_output_path) <> "/" <> id
   end
 
-  defp get_compose_config(job) do
+  @doc false
+  def get_compose_config(job) do
     Antidote.encode!(%{version: "3", services: build_services(job)})
```
Comment (PragTob) [inline lib/runner/job.ex:93]: so yeah see comments in tests but I generally like them better as private methods and then test it as part of where it is used (`start_job`) - making it public for testability should be the exception.

By making it public we couple our tests to their name, their parameter number and might also accidentally open them up for other people to use.

Keeping them private we only have one interface that is then usually more stable and can be abstracted better while we have the same confidence in it but more freedom to refactor its internals as long as we keep behaviour. I really love the chapter about this in Practical Object Oriented Design in Ruby by Sandi Metz

File: test/runner/job_test.exs [hunk 9]
```diff
@@ -45,4 +19,253 @@ defmodule ElixirBench.Runner.JobTest do
...
+      env_backup = Application.get_env(:runner, :benchmars_output_path)
+      Application.delete_env(:runner, :benchmars_output_path)
+
+      assert_raise ArgumentError, fn ->
+        Job.get_benchmars_output_path(job_fixture())
+      end
+
+      Application.put_env(:runner, :benchmars_output_path, env_backup)
+    end
+  end
```
Comment (PragTob) [inline test/runner/job_test.exs]: also think this should probably rather be tested as part of `start_job`

File: test/runner/job_test.exs [hunk 10]
```diff
@@ -45,4 +19,253 @@ defmodule ElixirBench.Runner.JobTest do
     assert job.context.worker_num_cores in 1..64
     assert job.context.worker_os in [:Linux, :Windows, :macOS]
   end
+
+  describe "start_job/6" do
```
Comment (PragTob) [inline test/runner/job_test.exs]: This isn't `start_job/6` anymore (parameter number has changed)

File: test/runner/job_test.exs [hunk 11]
```diff
@@ -45,4 +19,253 @@ defmodule ElixirBench.Runner.JobTest do
...
+
+      assert_raise MatchError, fn ->
+        Job.ensure_no_other_jobs!()
+      end
+
+      Process.exit(pid, :kill)
+    end
+  end
+
+  describe "get_benchmars_output_path/1" do
```
Comment (PragTob) [inline test/runner/job_test.exs]: this should also probably be `benchmarks` :)

File: test/runner/job_test.exs [hunk 12]
```diff
@@ -45,4 +19,253 @@ defmodule ElixirBench.Runner.JobTest do
...
+    end
+  end
+
+  describe "get_compose_config/1" do
+    test "return docker-compose setup for simple job" do
+      job = job_fixture()
+
+      compose_config =
+        Job.get_compose_config(job)
+        |> Antidote.decode!()
```
Comment (PragTob) [inline test/runner/job_test.exs]: usually one starts the pipe with a simple value (`job` in this case)

File: test/runner/job_test.exs [hunk 13]
```diff
@@ -45,4 +19,253 @@ defmodule ElixirBench.Runner.JobTest do
...
+
+      assert %{"runner" => %{"depends_on" => ["job_test_job_mysql", "job_test_job_postgres"]}} =
+               services
+
+      assert %{"runner" => %{"environment" => %{"PG_URL" => "postgres:postgres@localhost"}}} =
+               services
+
+      assert %{"runner" => %{"environment" => %{"MYSQL_URL" => "root@localhost"}}} = services
+    end
+  end
```
Comment (PragTob) [inline test/runner/job_test.exs]: I was unsure if your changes in `build_services` were ok in retaining behaviour, but these tests help my confidence - thank you :+1:

File: test/runner/job_test.exs [hunk 14]
```diff
@@ -45,4 +19,253 @@ defmodule ElixirBench.Runner.JobTest do
...
+    end
+
+    test "return empty map given invalid inputs" do
+      assert %{} = Job.format_measurement(measurement_fixture(), nil)
+      assert %{} = Job.format_measurement(%{}, "insert_mysql")
+
+      assert %{} = Job.format_measurement(nil, "insert_mysql")
+      assert %{} = Job.format_measurement(%{"a" => "b"}, "insert_mysql")
+    end
+  end
```
Comment (PragTob) [inline test/runner/job_test.exs:160]: where are we testing that we drop all that big measurement data? It's also not quite sure for me where that happens in the code as it's not that explicit, so I think we could improve the code and tests there :)

**Review by PragTob — COMMENTED** [review]
Sorry I somehow missed this - must have clicked away the notification or something :blush: . Comments inline

whatsapp image 2018-06-03 at 16 40 21

File: lib/runner/job.ex (hunk 3 above)
Comment (tallysmartins) [reply lib/runner/job.ex:48]: Yeah.. regarding the field names, I created an issue #11 to make them compatible later

File: test/runner/job_test.exs (hunk 6 above)
Comment (tallysmartins) [reply test/runner/job_test.exs]: in the line 41 I add another child process to the Supervisor that executes a sleep of 2 seconds, so when  Job.start_job spanws a new process to execute the benchmark it fails because there's already another child running.

File: test/runner/job_test.exs (hunk 7 above)
Comment (tallysmartins) [reply test/runner/job_test.exs]: I agree with you. But the job run is very complex and making methods public with `@doc false` sounds like a good solution in a way we can test the feature and corner cases without complex and time consuming tests, calling just `Job.start_job`. I got inspired in some discussions and posts like this one '[testing private functions](https://pragdave.me/blog/2017/03/31/tesing-private-functions.html)'  from PragDave, that uses a  library to make methods public under test environment. Another approach which I've found more interesting given its simplicity, is to make it public and add `@doc false` tags. Anyone can call it from the outside scope of the module but for me its a good signature and it also does not appear as an API/public method in the console, so I very liked it.

File: lib/runner.ex (hunk 1 above)
Comment (tallysmartins) [reply lib/runner.ex]: `revision |> fix()` :heavy_check_mark:

File: lib/runner/job.ex (hunk 4 above)
Comment (tallysmartins) [reply lib/runner/job.ex]: fixed! :heavy_check_mark:

File: lib/runner/job.ex (hunk 5 above)
Comment (tallysmartins) [reply lib/runner/job.ex]: Changed! :heavy_check_mark:

File: lib/runner/job.ex (hunk 8 above)
Comment (tallysmartins) [reply lib/runner/job.ex:93]: See my comment above... I think we have an exception here, `start_job` logic is very complex.

File: test/runner/job_test.exs (hunk 11 above)
Comment (tallysmartins) [reply test/runner/job_test.exs]: fixed! :heavy_check_mark:

File: test/runner/job_test.exs (hunk 13 above)
Comment (tallysmartins) [reply test/runner/job_test.exs]: I can make it more explicit and put the entire config within a pattern match like I did in `    test "return docker-compose setup for simple job" do`, wdyt?

File: test/runner/job_test.exs (hunk 14 above)
Comment (tallysmartins) [reply test/runner/job_test.exs:160]: there is the `format_measurements/2` I added a comment on it to explicit say that we are interested in just the `statistics data`

File: test/runner/job_test.exs (hunk 10 above)
Comment (tallysmartins) [reply test/runner/job_test.exs]: fixed!

File: test/runner/job_test.exs (hunk 12 above)
Comment (tallysmartins) [reply test/runner/job_test.exs]: fixed!

File: lib/runner/job.ex [hunk 15]
```diff
@@ -63,12 +55,11 @@ defmodule ElixirBench.Runner.Job do
     end
   end
 
-  defp ensure_no_other_jobs! do
+  def ensure_no_other_jobs! do
     %{active: 0} = Supervisor.count_children(ElixirBench.Runner.JobsSupervisor)
```
Comment (PragTob) [inline lib/runner/job.ex]: I still think we can easily keep this private and test as part of the `start_job` logic as it raises in the error case so no other logic is actually triggered.

File: lib/runner/job.ex [hunk 16]
```diff
@@ -93,11 +84,12 @@ defmodule ElixirBench.Runner.Job do
     end
   end
 
-  defp get_benchmars_output_path(%Job{id: id}) do
+  def get_benchmars_output_path(%Job{id: id}) do
```
Comment (PragTob) [inline lib/runner/job.ex]: should be called `brnchmarks` still with the k :)

File: lib/runner/job.ex [hunk 17]
```diff
@@ -177,6 +191,18 @@ defmodule ElixirBench.Runner.Job do
     }
   end
 
+  # Delete key that is not allowed in docker-compose otherwise it will break.
+  # This function iterates recursively over the services map and nested maps to delete the
+  # key for all services.
+  defp delete_non_docker_tag(services, tag) when is_map(services) do
```
Comment (PragTob) [inline lib/runner/job.ex:198]: Good name change :+1:

File: test/runner/job_test.exs [hunk 18]
```diff
@@ -45,4 +19,240 @@ defmodule ElixirBench.Runner.JobTest do
...
+      task_fun = fn job ->
+        {:ok, %{job | log: "success", status: 0}}
+      end
+
+      {:ok, job} = Job.start_job(job_fixture(), task_fun)
+
+      assert %{log: "success", status: 0} = job
+    end
+
+    test "not start job if there is other job running" do
```
Comment (PragTob) [inline test/runner/job_test.exs:34]: ah I see you already added the test here you just need to make the function private again I think

**Review by PragTob — COMMENTED** [review]
Looking good, just a couple things :clap:

File: lib/runner/job.ex (hunk 16 above)
Comment (tallysmartins) [reply lib/runner/job.ex]: :heavy_check_mark: changed

File: test/runner/job_test.exs (hunk 18 above)
Comment (tallysmartins) [reply test/runner/job_test.exs:34]: Yup

**Review by PragTob — APPROVED** [review]
img_20171221_144657_bokeh 1
